python/cudf_polars/cudf_polars/dsl/tracing.py

```diff
@@ -6,11 +6,174 @@
+) -> dict:  # pragma: no cover; requires CUDF_POLARS_LOG_TRACES=1
+    """
+    Collect statistics about the evaluation of an IR node.
+
+    Parameters
+    ----------
+    node_type
+        The type of the IR node.
+    frames
+        The list of DataFrames to capture information for. For ``phase="input"``,
+        this is typically the dataframes passed to ``IR.do_evaluate``. For
+        ``phase="output"``, this is typically the DataFrame returned from
+        ``IR.do_evaluate``.
+    extra
+        Extra information to log.
+    pid
+        The ID of the current process. Used for NVML memory usage.
+    device_handle
+        The pynvml device handle. Used for NVML memory usage.
+    phase
+        The phase of the evaluation. Either "input" or "output".
+    """
+    ir_name = node_type.__name__
+
+    d: dict[str, Any] = {
+        "type": ir_name,
+        f"count_frames_{phase}": len(frames),
+        f"frames_{phase}": [
+            {
+                "shape": frame.table.shape(),
+                "size": sum(col.device_buffer_size() for col in frame.table.columns()),
+            }
+            for frame in frames
+        ],
+    }
+    d[f"total_bytes_{phase}"] = sum(x["size"] for x in d[f"frames_{phase}"])
+
+    stats = rmm.statistics.get_statistics()
+    if stats:
+        d.update(
+            {
+                f"rmm_current_bytes_{phase}": stats.current_bytes,
+                f"rmm_current_count_{phase}": stats.current_count,
+                f"rmm_peak_bytes_{phase}": stats.peak_bytes,
+                f"rmm_peak_count_{phase}": stats.peak_count,
+                f"rmm_total_bytes_{phase}": stats.total_bytes,
+                f"rmm_total_count_{phase}": stats.total_count,
+            }
+        )
+
+    if extra:
+        d.update(extra)
+
+    if device_handle is not None:
+        processes = pynvml.nvmlDeviceGetComputeRunningProcesses(device_handle)
+        for proc in processes:
+            if proc.pid == pid:
+                d[f"nvml_current_bytes_{phase}"] = proc.usedGpuMemory
+                break
+
+    return d
+
+
+P = ParamSpec("P")
+
+
+def log_do_evaluate(
+    func: Callable[Concatenate[type[ir.IR], P], cudf_polars.containers.DataFrame],
+) -> Callable[Concatenate[type[ir.IR], P], cudf_polars.containers.DataFrame]:
+    """
+    Decorator for an ``IR.do_evaluate`` method that logs information before and after evaluation.
+
+    Parameters
+    ----------
+    func
+        The ``IR.do_evaluate`` method to wrap.
+    """
+    if not LOG_TRACES:
+        return func
+    else:  # pragma: no cover; requires CUDF_POLARS_LOG_TRACES=1
+
+        @functools.wraps(func)
+        def wrapper(
+            cls: type[ir.IR],
+            *args: P.args,
+            **kwargs: P.kwargs,
+        ) -> cudf_polars.containers.DataFrame:
+            # do this just once
+            pynvml.nvmlInit()
+            maybe_handle = get_device_handle()
+            pid = os.getpid()
```

Comment:
It feels like we don't have to call this for every `do_evaluate`. Can we cache it?

Reply:
Sure, that takes it from ~500ns to ~50ns for me locally.

I'll do the same for the device handle, which goes from ~5us to ~50ns.

Reply:
Gotcha. As of https://github.com/rapidsai/cudf/pull/19895/commits/be92e9f6a212e60992dc38bc2a0ede6057c74058 it's now cached.